File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_829_queue/Main.cpp
#include<iostream>
using namespace std;
const int N = 100010;
int q[N] , hh = 0 , tt = -1;

void push(int x){
    q[++tt] = x;
}

int query(){
    return q[hh];
}

void pop(){
    hh++;
}

int main(){
    string op;
    int x;
    int n;
    cin >> n;
    while(n--){
            cin >> op;
        if(op == "query"){
            cout << query() << endl;
        }else if( op == "push"){
            cin >> x;
            push(x);
        }else if( op == "pop"){
            pop();
        }else{
            string res =  hh <= tt ? "NO" : "YES";
            cout << res << endl;
        }   
    }
}

File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_829_queue/Main.java
package First_Level_0_Basic_Algorithm.Second_DataStructure.ds_829_queue;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static final int N = 100010;
    static int [] q = new int[N];
    static int hh = 0 , tt = -1;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.valueOf(br.readLine());
        while(n-- > 0){
            String [] op;
            op = br.readLine().split(" ");
            if(op[0].equals("query")){
                System.out.println(query());
            }else if(op[0].equals("push")){
                int x = Integer.valueOf(op[1]);
                push(x);
            }else if(op[0].equals("pop")){
                pop();
            }else{
                System.out.println(hh <= tt ? "NO" : "YES");
            }
        }
        br.close();
    }

    public static void push(int x){
        q[++tt] = x;
    }

    public static int query(){
        return q[hh];
    }

    public static void pop(){
        hh++;
    }
}
